shared-libs/lineage/src/lineage.js

```diff
@@ -0,0 +1,403 @@
+      if (contactDoc) {
+        doc.contact = contactDoc;
+      }
+      current = doc.contact;
+    }
+
+    while (current) {
+      if (current.parent && current.parent._id) {
+        var parentDoc = findById(current.parent._id, parents);
+        if (parentDoc) {
+          current.parent = parentDoc;
+        }
+      }
+      current = current.parent;
+    }
+  });
+  return docs;
+};
+
+var hydrateLeafContacts = function(docs, contacts) {
+  var subDocsToHydrate = [];
+  docs.forEach(function(doc) {
+    var current = doc;
+    if (doc.type === 'data_record') {
+      current = doc.contact;
+    }
+    while (current) {
+      subDocsToHydrate.push(current);
+      current = current.parent;
+    }
+  });
+  fillContactsInDocs(subDocsToHydrate, contacts);
+  return docs;
+};
+
+var hydratePatient = function(doc) {
+  return fetchPatientLineage(doc).then(function(patientLineage) {
+    if (patientLineage.length) {
+      var patientDoc = patientLineage.shift();
+      fillParentsInDocs(patientDoc, patientLineage);
+      doc.patient = patientDoc;
+    }
+    return doc;
+  });
+};
+
+var hydratePatients = function(docs) {
+  return Promise.all(docs.map(hydratePatient)).then(function() {
+    return docs;
+  });
+};
+
+var hydrateDocs = function(docs) {
+  if (!docs.length) {
+    return Promise.resolve([]);
+  }
+  var parentIds = collectParentIds(docs);
+  var hydratedDocs = JSON.parse(JSON.stringify(docs));
+  return fetchDocs(parentIds)
+    .then(function(parents) {
+      hydrateParents(hydratedDocs, parents);
+      return fetchDocs(collectLeafContactIds(hydratedDocs));
+    })
+    .then(function(contacts) {
+      hydrateLeafContacts(hydratedDocs, contacts);
+      return hydratePatients(hydratedDocs);
+    });
+};
+
+// Minifies things you would attach to another doc:
+//   doc.parent = minify(doc.parent)
+// Not:
+//   minify(doc)
+var minifyLineage = function(parent) {
+  if (!parent || !parent._id) {
+    return parent;
+  }
+
+  var result = { _id: parent._id };
+  var minified = result;
+  while (parent.parent && parent.parent._id) {
+    minified.parent = { _id: parent.parent._id };
+    minified = minified.parent;
+    parent = parent.parent;
+  }
+  return result;
+};
+
+module.exports = {
+  init: function(dependencies) {
+    dependencies = dependencies || {};
+    Promise = dependencies.Promise;
+    DB = dependencies.DB;
+    return module.exports;
+  },
+
+  /**
+   * Given a doc id get a doc and all parents, contact (and parents) and
+   * patient (and parents)
+   * @param id The id of the doc
+   * @returns Promise
+   */
+  fetchHydratedDoc: fetchHydratedDoc,
+
+  /**
+   * Given an array of minified docs bind the parents and contacts
+   * @param docs The array of docs to hydrate
+   * @returns Promise
+   */
+  hydrateDocs: hydrateDocs,
+
+  /**
+   * Remove all hyrdrated items and leave just the ids
+   * @param doc The doc to minify
+   */
+  minify: function(doc) {
```

Comment:
Maybe we need to call this `minifyDoc` to distinguish it from `minifyLineage` below?

Reply:
Yeah we could look at that kind of thing. It's sort of a gross api at first cut because of all the extra functions that were needed below. A lot of this could come from really refactoring how webapp interacts with lineage, because it interacts with it in a very different way, which is why there are similar functions.